File: node/examples/webhook-server.ts
/**
 * Пример: приём вебхука H2H с проверкой подписи на встроенном `node:http`.
 *
 * Ключевой момент: подпись проверяется по СЫРОМУ телу запроса, поэтому
 * тело читается как есть, до `JSON.parse` — пересборка JSON меняет
 * порядок ключей и ломает подпись.
 *
 * Запуск:
 *   WATA_ACQUIRING_TOKEN=<jwt> node --experimental-strip-types examples/webhook-server.ts
 */

import { createServer } from 'node:http';
import { WataClient } from '../src/index.js';

const token = process.env.WATA_ACQUIRING_TOKEN;
if (!token) {
  throw new Error('Задайте переменную окружения WATA_ACQUIRING_TOKEN.');
}

const client = new WataClient({ acquiring: token, environment: 'production' });

const server = createServer((req, res) => {
  if (req.method !== 'POST') {
    res.writeHead(405).end();
    return;
  }

  const chunks: Buffer[] = [];
  req.on('data', (chunk: Buffer) => chunks.push(chunk));
  req.on('end', () => {
    void (async () => {
      const rawBody = Buffer.concat(chunks); // именно сырые байты, не JSON.parse
      const signature = req.headers['x-signature'];

      if (typeof signature !== 'string') {
        res.writeHead(400).end('Отсутствует заголовок X-Signature');
        return;
      }

      try {
        const isValid = await client.webhooks.verify(rawBody, signature);
        if (!isValid) {
          res.writeHead(401).end('Неверная подпись');
          return;
        }

        const event = client.webhooks.parse(rawBody);
        console.log('Получено событие вебхука:', event.kind, event.transactionStatus, event.orderId);

        // Обработчик обязан быть идемпотентным (возможны повторные доставки
        // до 32 часов для постоплатных/возвратных вебхуков) и вернуть 200.
        res.writeHead(200).end('OK');
      } catch (err) {
        console.error('Не удалось обработать вебхук:', err);
        res.writeHead(500).end();
      }
    })();
  });
});

server.listen(3000, () => {
  console.log('Слушаю вебхуки WATA на http://localhost:3000');
});

File: python/examples/webhook_server.py
"""Пример: приём и проверка подписи вебхука WATA (только стандартная библиотека).

Запуск::

    pip install wata[webhooks]
    WATA_ACQUIRING_TOKEN=<jwt терминала эквайринга> python examples/webhook_server.py

Затем отправьте POST на http://localhost:8000/webhooks/wata с заголовком
`X-Signature` и телом события — SDK проверит подпись SHA512withRSA по ключу,
полученному через `GET /public-key`, и разберёт событие.
"""

from __future__ import annotations

import os
from http.server import BaseHTTPRequestHandler, HTTPServer

from wata import WataClient, WataWebhookError

client = WataClient(acquiring=os.environ.get("WATA_ACQUIRING_TOKEN", ""))


class WebhookHandler(BaseHTTPRequestHandler):
    def do_POST(self) -> None:  # noqa: N802 — имя метода задано http.server
        content_length = int(self.headers.get("Content-Length", 0))
        raw_body = self.rfile.read(content_length)  # сырое тело: bytes, не разобранный JSON
        signature = self.headers.get("X-Signature", "")

        try:
            is_valid = client.verify_webhook(raw_body, signature)
        except WataWebhookError as exc:
            print("Не удалось проверить подпись:", exc)
            self.send_response(400)
            self.end_headers()
            return

        if not is_valid:
            print("Подпись не совпала — запрос отклонён")
            self.send_response(401)
            self.end_headers()
            return

        event = client.parse_webhook(raw_body)
        print(
            f"Событие {event.kind}: заказ={event.order_id} статус={event.transaction_status} "
            f"сумма={event.amount} {event.currency}"
        )

        # Обработчик обязан быть идемпотентным: в реальном приложении здесь
        # нужно свериться с id транзакции/события в своей БД перед повторной
        # обработкой, и в любом случае вернуть 200 — иначе WATA будет
        # повторять доставку постоплатных/возвратных вебхуков до 32 часов.

        self.send_response(200)
        self.end_headers()

    def log_message(self, format: str, *args: object) -> None:  # noqa: A002 — сигнатура базового класса
        pass  # приглушаем стандартный лог http.server, у нас свой print выше


def main() -> None:
    server = HTTPServer(("0.0.0.0", 8000), WebhookHandler)
    print("Слушаю вебхуки WATA на http://0.0.0.0:8000/webhooks/wata ...")
    try:
        server.serve_forever()
    finally:
        client.close()


if __name__ == "__main__":
    main()
